certchain/certmanager/multicert_disk_cache.go

```diff
@@ -0,0 +1,181 @@
+	"errors"
+	"path"
+
+	"github.com/gofrs/flock"
+	"github.com/google/renameio"
+	"github.com/google/webpackager/certchain"
+	"github.com/google/webpackager/certchain/certchainutil"
+	"github.com/hashicorp/go-multierror"
+)
+
+var (
+	errEmptyCertDir      = errors.New("certmanager: empty CertDir")
+	errEmptyCertFilename = errors.New("certmanager: empty CertFilename")
+	errEmptyOCSPFilename = errors.New("certmanager: empty OCSPFilename")
+)
+
+// MultiCertDiskCache is a Cache on a local filesystem. It writes the
+// certificate chain in the PEM format and the OCSP response in the DER format
+// to separate files as specified by MultiCertDiskCacheConfig. It uses the
+// digest of the certificate chain as the base filename for the cert and OCSP
+// files.
+type MultiCertDiskCache struct {
+	MultiCertDiskCacheConfig
+}
+
+var _ Cache = (*MultiCertDiskCache)(nil)
+
+// MultiCertDiskCacheConfig configures DiskCache.
+type MultiCertDiskCacheConfig struct {
+	// CertDir locates the directory to write the certificate chain to.
+	// If CertDir is empty, the certificate chain is not cached.
+	CertDir string
+
+	// LatestCertFilename specifies the filename to be used for the latest
+	// version of the certificate. The file will be located in CertDir.
+	LatestCertFilename string
+
+	// LatestOCSPFilename specifies the filename to be used for the latest
+	// version of the OCSP. The file will be located in CertDir.
+	LatestOCSPFilename string
+
+	// LockFilename locates the lock file. Must be non-empty. The file will be
+	// located in CertDir.
+	LockFilename string
+}
+
+// NewMultiCertDiskCache creates and initializes a new MultiCertDiskCache.
+func NewMultiCertDiskCache(config MultiCertDiskCacheConfig) *MultiCertDiskCache {
+	return &MultiCertDiskCache{config}
+}
+
+// Read reads the certificate chain and the OCSP response from local files
+// and reproduces an AugmentedChain. Read uses digest to retrieve the
+// AugmentedChain for that digest.  If digest is an empty string, Read returns
+// the latest revision of the AugmentedChain as referenced by LatestCertFilename
+// and LatestOCSPFilename.  Read only works only when d.CertDir is non-empty and
+// for cases where digest is an empty string, it only works when
+// d.LatestCertFilename and d.LatestOCSPFilename are non-empty,
+// otherwise returns an error. Read returns a multierror.Error
+// (hashicorp/go-multierror) to report as many problems as possible.
+func (d *MultiCertDiskCache) Read(digest string) (*certchain.AugmentedChain, error) {
+	var errs *multierror.Error
+
+	if d.CertDir == "" {
+		errs = multierror.Append(errs, errEmptyCertDir)
+	}
+
+	if err := errs.ErrorOrNil(); err != nil {
+		return nil, err
+	}
+
+	lock := flock.New(path.Join(d.CertDir, d.LockFilename))
+	errs = multierror.Append(errs, lock.RLock())
+
+	if err := errs.ErrorOrNil(); err != nil {
+		return nil, err
+	}
+
+	cert := path.Join(d.CertDir, digest+".pem")
+	ocsp := path.Join(d.CertDir, digest+".ocsp")
+	if digest == "" {
```

Comment:
This kind of behavior should be documented in GoDoc of the `Cache` interface, not (just) GoDoc of this method. Also, I think it's caller-friendly to define a separate method for this purpose rather than giving empty string a special meaning. Compare:

```go
ac, err := cache.Read("")
    // versus
ac, err := cache.ReadLatest()
```

Reply:
Added ReadLatest method to Cache and all structs affected.